### ISSUE TITLE
BatchNorm doesn't use CuDNN for backwards in evaluation mode

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#4284]
In principle we should be able to use the CuDNN kernel by passing nullptr for save_mean and save_std. However, when I do this, gradcheck fails.

```
diff --git a/aten/src/ATen/native/cudnn/BatchNorm.cpp b/aten/src/ATen/native/cudnn/BatchNorm.cpp
index 14730ee..ab8690e 100644
--- a/aten/src/ATen/native/cudnn/BatchNorm.cpp
+++ b/aten/src/ATen/native/cudnn/BatchNorm.cpp
@@ -146,7 +146,10 @@ std::tuple<Tensor, Tensor, Tensor> cudnn_batch_norm_backward(
   CheckedFrom c = "cudnn_batch_norm_backward";
   setCuDNNStreamToCurrent();
 
-  checkAllDefined(c, {input, grad_output, weight, save_mean, save_var});
+  checkAllDefined(c, {input, grad_output, weight});
+  if (training) {
+    checkAllDefined(c, {save_mean, save_var});
+  }
   checkAllSameGPU(c, {input, grad_output, weight, save_mean, save_var});
   if (input->type().scalarType() == ScalarType::Half) {
     checkScalarType(c, weight, ScalarType::Float);
@@ -160,8 +163,11 @@ std::tuple<Tensor, Tensor, Tensor> cudnn_batch_norm_backward(
   checkDimRange(c, input, 2, 6 /* exclusive */);
   checkSameSize(c, input, grad_output);
   auto num_features = input->size(1);
-  for (auto t : {weight, save_mean, save_var}) {
-    checkNumel(c, t, num_features);
+  checkNumel(c, weight, num_features);
+  if (training) {
+    for (auto t : {save_mean, save_var}) {
+      checkNumel(c, t, num_features);
+    }
   }
 
   cudnnBatchNormMode_t mode;
@@ -197,8 +203,8 @@ std::tuple<Tensor, Tensor, Tensor> cudnn_batch_norm_backward(
     grad_weight_t.data_ptr(),
     grad_bias_t.data_ptr(),
     epsilon,
-    save_mean->data_ptr(),
-    save_var->data_ptr()));
+    training ? save_mean->data_ptr() : nullptr,
+    training ? save_var->data_ptr() : nullptr));
 
   return std::tuple<Tensor,Tensor,Tensor>{grad_input_t, grad_weight_t, grad_bias_t};
 }
zation.cpp a/torch/csrc/autograd/functions/batch_normalization.cpp b/torch/csrc/autograd/functions/batch_normal 
index d01f755..833d643 100644
--- a/torch/csrc/autograd/functions/batch_normalization.cpp
+++ b/torch/csrc/autograd/func
                              tions/batch_normalization.cpp
@@ -107,7 +107,7 @@ auto BatchNormBackward::apply(const variable_list& grad_outputs) -> variable_lis
   use_cudnn = (input.type().backend() == at::kCUDA
                && (input.type().scalarType() != at::kHalf
                || weight.type().scalarType() == at::kFloat)
-               && weight.defined() && bias.defined() && training
+               && weight.defined() && bias.defined()
                && input.size(0) <= 131070
                && cudnn_enabled && CUDNN_VERSION >= 5110L);
 #endif
```

I also tried tracking down the history but the `training` special case seems to have been here since the very beginning of Pytorch.

CC @ngimel

**Comment by ngimel** [comment]
No, cudnn can not be used for backward in evaluation mode, because math for backward in evaluation mode is different (because math for forward in evaluation mode is different), and cudnn does not support it. 
Generally, no one should be ever running bn in evaluation mode (be it forward or backward), for eval mode it should be merged into previous convolution, there was an early pr to that effect that was abandoned https://github.com/pytorch/pytorch/pull/901

**Comment by ezyang** [comment]
Thanks! Closing this.
